sdks/java/harness/src/main/java/org/apache/beam/fn/harness/MapRunner.java

```diff
@@ -0,0 +1,102 @@
+ * limitations under the License.
+ */
+
+package org.apache.beam.fn.harness;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Multimap;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import org.apache.beam.fn.harness.data.BeamFnDataClient;
+import org.apache.beam.fn.harness.data.MultiplexingFnDataReceiver;
+import org.apache.beam.fn.harness.fn.ThrowingFunction;
+import org.apache.beam.fn.harness.fn.ThrowingRunnable;
+import org.apache.beam.fn.harness.state.BeamFnStateClient;
+import org.apache.beam.model.pipeline.v1.RunnerApi;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PCollection;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PTransform;
+import org.apache.beam.sdk.fn.data.FnDataReceiver;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.util.WindowedValue;
+
+/**
+ * Add an abstraction which makes it easy to create runners which can wrap simple map functions
+ * which are typical in a lot of transforms.
+ *
+ * <p>TODO: Add support for DoFns which are actually user supplied map/lambda functions instead
+ * of using the FnApiDoFnRunner instance.
+ */
+public class MapRunner<InputT, OutputT> {
+  /** A factory for {@link BeamFnDataReadRunner}s. */
+  public abstract static class Factory<InputT, OutputT>
+      implements PTransformRunnerFactory<MapRunner<InputT, OutputT>> {
+
+    @Override
+    public final MapRunner<InputT, OutputT> createRunnerForPTransform(
+        PipelineOptions pipelineOptions,
+        BeamFnDataClient beamFnDataClient,
+        BeamFnStateClient beamFnStateClient,
+        String pTransformId,
+        PTransform pTransform,
+        Supplier<String> processBundleInstructionId,
+        Map<String, PCollection> pCollections,
+        Map<String, RunnerApi.Coder> coders,
+        Map<String, RunnerApi.WindowingStrategy> windowingStrategies,
+        Multimap<String, FnDataReceiver<WindowedValue<?>>> pCollectionIdsToConsumers,
+        Consumer<ThrowingRunnable> addStartFunction,
+        Consumer<ThrowingRunnable> addFinishFunction) throws IOException {
+
+      Collection<FnDataReceiver<WindowedValue<OutputT>>> consumers =
+          (Collection) pCollectionIdsToConsumers.get(
+              getOnlyElement(pTransform.getOutputsMap().values()));
+
+      MapRunner<InputT, OutputT> runner = new MapRunner<>(
+          createMapFunctionForPTransform(pTransformId, pTransform),
+          MultiplexingFnDataReceiver.forConsumers(consumers));
+
+      pCollectionIdsToConsumers.put(
+          Iterables.getOnlyElement(pTransform.getInputsMap().values()),
+          (FnDataReceiver) (FnDataReceiver<WindowedValue<InputT>>) runner::map);
+      return runner;
+    }
+
+    protected abstract ThrowingFunction<InputT, OutputT> createMapFunctionForPTransform(
```

Comment:
Is the motivation of having an abstract function that returns a function (rather than simply make the mapping function itself abstract) simply to be able to lazily parse the proto (vs. do it in the constructor)?

Reply:
There is no lazy parsing, it prevents those factories from knowing what type they are constructing/returning and hides the fact that the output is being sent to a FnDataReceiver.